### DISCUSSION TITLE
[HOWTO] 40-pin GPIO & ADLX345 on CB1

### DISCUSSION THREAD

**Discussion by Michelinu** [bigtreetech/CB1#47] [Show and tell]
Hello,
I started this new thread in order to regroup all the info and made them are easier to source.

### **40-pin connector / GPIO**
As somebody already discovered, the CB1 uses a **custom IO pinout** for the 40-pin connector. All the IO "naming" that works on standard Raspberry doesn't work on CB1. I discovered a way to get some working IO pins.

Here it's the pinout (from the official Manta M4P user manual)
![](https://user-images.githubusercontent.com/5544539/193459421-78ff9651-e8a3-4785-a75b-11f206eae7e0.png)
You have to take the name after the underscore for the PIN you need. Then use this table https://wiki.pine64.org/wiki/SOPine_GPIO_Alternate_Functions_Table to get the `GPIO<number>`.
I tried with some "PC" pins and it seems to work.

**Example**: pin n.15 -> `GPIO22_PC10` -> `PC10` -> `GPIO74`
If you are having some troubles getting GPIO working under **Moonraker**, check this issue: https://github.com/bigtreetech/CB1/issues/44
###
###
###
### **ADXL345**
**stas2z** already provided the instructions (https://github.com/bigtreetech/CB1/discussions/9#discussioncomment-3540300). I did a new more comprehensible wiring scheme.
![](https://user-images.githubusercontent.com/5544539/193461714-fea2ce27-637d-4b8d-a04b-e35f6ca2c67b.png)

It works also with **official BTT images** (I tried with latest v2.2.0 release). If you flash the "Klipper" image all the required **libraries**, **NumPy** and the **linux MCU** host are already installed. If you are using the "minimal" image you have to install them by following the Klipper documentation: https://www.klipper3d.org/Measuring_Resonances.html

Since the `raspi-config` command is not available, I added the following entries in my `/boot/BoardEnv.txt` to **enable SPI**:
```
overlays=spi-spidev
param_spidev_spi_bus=1
param_spidev_spi_cs=1
param_spidev_max_freq=1000000
```

Here's my current `adxl345.cfg` file (that you can #include in your `printer.cfg` or paste the content directly in that):
```
[mcu rpi]
serial: /tmp/klipper_host_mcu

[adxl345]
cs_pin: rpi:gpio74
spi_bus: spidev1.1

[resonance_tester]
accel_chip: adxl345
probe_points:
  100, 100, 20
```

**Comment by BCTANBC** [comment]
Work very well with my M8P with CB1. Thanks for the afford.

**Comment by Agent-047185** [comment]
hi guys, thanks for the great infos! planning to use a blue adxl like this one with a manta m4p+cb1.I have one question: this shows a diagram to the Pi compute module. Does anyone knows where to find a diagram to wire the ADXL on the Manta GPIO please? sorry noob here ^^.

**Reply by freakydude** [reply]
Should be exactly the same...

**Reply by Pneumanifest** [reply]
> hi guys, thanks for the great infos! planning to use a blue adxl like this one with a manta m4p+cb1.I have one question: this shows a diagram to the Pi compute module. Does anyone knows where to find a diagram to wire the ADXL on the Manta GPIO please? sorry noob here ^^.

Manta has onboard spi port

[adxl345]
cs_pin: PD9
spi_bus: spi1
axes_map: z,y,-x

[resonance_tester]
accel_chip: adxl345
probe_points:
    100, 100, 20  # an example

**Reply by Agent-047185** [reply]
@Pneumanifest thank you! Gonna try this soon

**Comment by joshuanazareth97** [comment]
Hey, thanks for all the info! Does anyone know if any of the CB1's GPIO pins are PWM? I'd like to control a fan and some LEDs using mosfets triggered by GPIO pins.

**Reply by eugenmeltis3pg** [reply]
Hello! Did you manage to add a led with MOSFET and connect it to a CB1 GPIO pin?

**Comment by jarvis710** [comment]
hello; First, forgive Google for the translation and second, thanks for what you published. It worked for me the first time.
Now, I have a board mantae3ez+cb1. as I commented it works; but what if i wanted to connect a second adxl345 to configure one for the hotend and one for the bed.
I'm more than new to this, but curious. If you could help me with this information to take advantage of a second adxl345 that I have and not waste it.
Thanks in advance and greetings

**Comment by fabio-s-franco** [comment]
No luck for me. Got CB1 + manta m8p v1.1 + tft35 on spi. 

error is:
```
Invalid adxl345 id (got ff vs e5).
```

BoardEnv.txt
```
overlays=tft35_spi spidev0_0
param_spidev0_0_spi_bus=0
param_spidev0_0_spi_cs=0
param_spidev0_0_max_freq=1000000
```

on readme says to use 1_2 (also tried that) due to the screen on 1.1.
But no difference.

Plugged it on pin 26 which according to readme, maps to GPIO200:

```
[adxl345]
cs_pin: host: gpiochip0/gpio200 
spi_bus: spidev0.0
axes_map: -y, x, -z

[resonance_tester]
accel_chip: adxl345
probe_points:
    115, 115, 20
```

Tried many other combinations.  Also tried the onboard SPI port which causes CB1 not to boot. The moment I remove the physical connection it resumes and boots. Really tried everything on the onboard port. Software and hardware SPI.

I am 2 weeks trying a way for this to work.
Yesterday decided to try on an arduino uno.
Took me 15 minutes to get it working on SPI. This is becoming insane.

This is what I got:

<img width="661" alt="image" src="https://github.com/bigtreetech/CB1/assets/37698530/ffd4df32-4070-4c81-ab98-ca87507316ee">

It is 3.3v (not 5v tolerant).

gpioinfo:

```
...
        line  69:      unnamed       unused   input  active-high
        line  70:      unnamed       unused   input  active-high
        line  71:      unnamed   "spi1 CS1"  output   active-low [used]
        line  72:      unnamed       unused   input  active-high
        line  73:      unnamed       unused   input  active-high
        line  74:      unnamed   "i2c-gpio"  output  active-high [used open-drain]
        line  75:      unnamed   "spi1 CS0"  output   active-low [used]
        line  76:      unnamed   "i2c-gpio"  output  active-high [used open-drain]
        line  77:      unnamed       unused   input  active-high
        line  78:      unnamed         "dc"  output  active-high [used]
...
        line 200:      unnamed    "klipper"  output  active-high [used]
```

Out of ideas. This board has been so hard to work with. Documentation is very shallow and info spread everywhere, I am considering returning it and get something else. It has been a time pit to get even simple things work due to lack of  detailed documentatiom.

**Reply by freakydude** [reply]
- BoardEnv.txt
  - no spi-dev overlay enabled
- printer.cfg
  - ```
    [adxl345]
    cs_pin: PC4
    ```
- Manta M8P
  - ![IMG20230613175722-adxl](https://github.com/bigtreetech/CB1/assets/32776748/2f9b3814-88f9-45ea-baa6-15a6c7e9ca38)
  - ![IMG20230613174611-adxl](https://github.com/bigtreetech/CB1/assets/32776748/9e43373c-fdc9-4143-80fe-54274413117c)

Sorry for my different cable colors ... 
Match sensor colors from sensor picture -> manta picture
Brown -> Black
Blue -> Green
Green -> Brown
Yellow -> Blue
Orange -> Yellow
Red -> Red




With best regards,


freakyDude

**Reply by fabio-s-franco** [reply]
Thanks freakyDude. Unfortunately won't work for me because I need overlay. My Hotend use PT100 with max amplifier :(

**Reply by freakydude** [reply]
Thanks 

Maybe the tft with overlay anyway .. What I was trying to say was, you don't need to configure a overlay for the adxl as you tried in your posted example config above

You may need the overlay for you tft.. But also in this case, isn't it connected to the exp pins from the manta instead of the gpio?

**Reply by anton-vinogradov** [reply]
@fabio-s-franco,
https://www.reddit.com/r/BIGTREETECH/comments/14m222t/btt_pi_and_btt_adxl345_settings/

**Comment by fabio-s-franco** [comment]
No, it's connected via SPI port it has via flat cable connector in the back
of the board. Judging by the gpioinfo output, it is routing through CB1
GPIO pins.
It is on SPI1 bus, which is all the the CB1 provides. To connect through
it, I must use the overlay otherwise the screen doesn't work.
I don't know where SPI0 are for the stepper drivers and SPI2 is the MCU
only available SPI port. This SPI2 port doesn't work and I am not sure why.
The MCU simply won't boot. My assumption is that the board itself is bugged.

**Comment by jarvis710** [comment]
Googling and thanks to this post, I was able to activate the two adxl in my mantae3ez, one for the spi and the other for the mcu. in case it is useful to someone

[mcu adxl]
serial: /tmp/klipper_host_mcu

[adxl345 hotend]
cs_pin: adxl:gpio74
spi_bus: spidev1.1

[adxl345 bed]
cs_pin:PC15
spi_software_miso_pin: PC11
spi_software_mosi_pin: PC12
spi_software_sclk_pin: PC10


[resonance_tester]
accel_chip_x: adxl345 hotend
accel_chip_y: adxl345 bed
max_smoothing: 0.090
probe_points:
  117.5, 117.5, 20

boardenv.txt

overlays=spi-spidev
param_spidev_spi_bus=1
param_spidev_spi_cs=1
param_spidev_max_freq=1000000

SPI1_CLK=PH6
SPI1_MISO=PH8
SPI1_MOSI=PH7
MCP2515_CS=PC11
MCP2515_IRQ=PC9


![20230616_014756](https://github.com/bigtreetech/CB1/assets/29159363/9d178f3e-96ac-474b-8c92-4bd37d10a028)

![1-12_650x](https://github.com/bigtreetech/CB1/assets/29159363/e67d52ae-68db-413e-9536-30ef3e65efbe)

**Comment by ReynosoRolandJosep** [comment]
Will this work on a BTT Pi V1.2 with TFT35 SPI LCD?

**Reply by anttix** [reply]
Certainly not. SPI1 pins are shared with the TFT35 connector.

One can try to wire ADLX345 to 3 (CS), 5 (SCLK), 27 (MOSI/SDA), 29 (MISO/SDO) on the 40 pin connector.
These connect to exactly the same pins that the "special" ADLX345 connector does, so instructions for configuring it should apply. YMMV.

**Comment by fabio-s-franco** [comment]
For those still struggling with manta, forget it. SPI doesn't work. The only way I made it work was through CB1.

Before I could post the solution, I burned my board and lost my cfg file. But I eventually found this article, which takes a very similar approach:

https://www.3docity.com.au/blogs/3dprinting/adxl345-and-bigtreetech-cb1-input-shaper-setup
